src/rust/engine/fs/store/src/lib.rs

```diff
@@ -804,13 +831,7 @@ impl Store {
               .mode(if is_executable { 0o755 } else { 0o644 })
               .open(&destination)
           })
-          .and_then(|mut f| {
-            f.write_all(&bytes)?;
-            // See `materialize_directory`, but we fundamentally materialize files for other
-            // processes to read; as such, we must ensure data is flushed to disk and visible
-            // to them as opposed to just our process.
-            f.sync_all()
```

Comment:
I don't follow why this is safe to remove?

Reply:
Assuming no one calls `materialize_file` other than as part of `materialize_directory`, this isn't being removed, just moved :) `fsync` is an instruction to the OS to say "whatever writes may have buffered for this file, make sure they're actually flushed to disk" (it doesn't refer to any particular writes, it refers to "all pending writes"). So _where_ we `fsync` isn't super important semantics-wise.

The reason this is a bunch faster is that most filesystems these days operate roughly as an append-able transaction log, which isn't per-file, so `fsync` tends to be implemented on the OS level not as "flush all pending changes for this file" but "flush all pending changes to all files". What this change effectively does is switch from doing 100 `fsync`s, each of which is slow (because each has new data to flush to disk), to doing 1 `fsync` which is slow, and 99 `fsync`s which the OS notices are no-ops (because there are no pending changes) so skips. It's heavily reliant on OS behaviour as an optimisation, but is totally correct in terms of correctness because we're still calling `fsync` for each file :)

Reply:
Thank you for the explanation!

Reply:
> Assuming no one calls materialize_file other than as part of materialize_directory

Mm. Good point. That would be an unfortunate thing to discover.